aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctions.cpp

```diff
@@ -101,5 +102,102 @@ Tensor nested_from_padded_cuda(
     return at::native::nested_from_padded_generic(padded, sizes);
   }
 }
+
+Tensor _collapse_two_dims_3(const Tensor& input, int64_t dim1, int64_t dim2) {
+  TORCH_CHECK(dim1 > 0, "dim1: Cannot collapse dim 0.");
+  TORCH_CHECK(dim2 > 0, "dim2: Cannot collapse dim 0.");
+  TORCH_CHECK(dim2 - 1 == dim1, "dim2 must be one more than dim1.")
+  TORCH_CHECK(dim1 == 1, "dim1 must be 1.")
+  TORCH_CHECK(input.dim() == 3, "Expected input to be 3 dim.");
+
+  auto* nt_input = get_nested_tensor_impl(input);
+  TORCH_CHECK(nested_tensor_impl_is_contiguous(nt_input));
+  Tensor nt_sizes = nt_input->get_nested_size_tensor();
+
+  Tensor sizes_dim1 = at::native::narrow(nt_sizes, 1, 0, 1);
+  Tensor sizes_dim2 = at::native::narrow(nt_sizes, 1, 1, 1);
+  Tensor new_nt_sizes = sizes_dim1 * sizes_dim2;
+
+  Tensor result = at::detail::make_tensor<NestedTensorImpl>(nt_input->get_buffer(), new_nt_sizes);
+  TORCH_CHECK(result.dim() == 2, "Expected result to be 2 dimensional.");
+  return result;
+}
+
+Tensor batch_offsets_from_efficient_size(const Tensor& ef_sizes) {
+  int64_t* nt_sizes_ptr = ef_sizes.data_ptr<int64_t>();
+  int64_t ef_sizes_size_0 = ef_sizes.sizes()[0];
+  Tensor offsets = at::empty({1 + ef_sizes_size_0}, at::kLong);
+  int64_t* offsets_ptr = offsets.data_ptr<int64_t>();
+  offsets_ptr[0] = 0;
+  int64_t ef_sizes_size_1 = ef_sizes.sizes()[1];
+  for (const auto i : c10::irange(ef_sizes_size_0)) {
+    int64_t prod = 1;
+    for (const auto j : c10::irange(ef_sizes_size_1)) {
+      prod = prod * nt_sizes_ptr[i * ef_sizes_size_1 + j];
+    }
+    offsets_ptr[i + 1] = offsets_ptr[i] + prod;
+  }
+  return offsets;
+}
+
+Tensor NestedTensor_to_padded_tensor_cuda(const Tensor& t, double padding) {
+  int64_t t_dim = t.dim();
+  if ((t_dim >= 2 && t_dim <= 4)) {
+    auto* nt_input = get_nested_tensor_impl(t);
+    TORCH_CHECK(nested_tensor_impl_is_contiguous(nt_input));
+    const auto& nt_buffer = nt_input->get_buffer();
+    const auto nt_input_opt_size_2 = nt_input->opt_size(2);
+
+    if (t_dim == 3 && nt_input_opt_size_2) {
+      Tensor output = NestedTensor_to_padded_tensor_cuda(
+          _collapse_two_dims_3(t, 1, 2), padding);
+      return output.reshape({output.sizes()[0], -1, *nt_input_opt_size_2});
+    }
+
+    Tensor nt_sizes = nt_input->get_nested_size_tensor();
+    Tensor offsets = batch_offsets_from_efficient_size(nt_sizes);
+    auto new_size = NestedTensor_get_max_size(*nt_input);
+    new_size.insert(new_size.begin(), nt_sizes.sizes()[0]);
+    Tensor output = at::empty(IntArrayRef(new_size), nt_buffer.options());
+
+    int64_t input_dim = nt_sizes.sizes()[1];
+    int64_t batch_size = nt_sizes.sizes()[0];
+    // TODO: Remove need for cat here
+    at::Tensor metadata = at::cat({offsets, nt_sizes.reshape(-1)});
+    metadata = metadata.to(at::Device(kCUDA), at::kInt, true, true);
```

Comment:
Doing async transfer here is dangerous because metadata may go out of scope before actual transfer happens